File: sdks/python/tests/test_preemptive_timeout.py
"""
Regression test for preemptive flow timeout (Python engine).

Mirrors sdks/nodejs/tests/test.preemptive-timeout.js. Before, the flow timeout
only set a flag checked between nodes, so a hung node parked the launch loop.
After, run() sets an abort asyncio.Event on timeout and _race_abort cancels the
in-flight node task (cancellation propagates into httpx/SDK calls), so a hung
node is cut loose at ~timeout. Normal runs are unaffected.

Deterministic & offline: we stub a node's process to hang — no network, no keys.

Run with: python tests/test_preemptive_timeout.py
"""

from __future__ import annotations

import asyncio
import sys
import time
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parents[1]))

from src.engine import Zv1  # noqa: E402

FLOW = {
    "nodes": [
        {"id": "in", "type": "input-data", "settings": {"key": "x"}},
        {"id": "out", "type": "output-data", "settings": {"key": "out"}},
    ],
    "links": [
        {"from": {"node_id": "in", "port_name": "value"},
         "to": {"node_id": "out", "port_name": "value"}},
    ],
}


async def test_normal_run_unaffected() -> None:
    print("\n--- normal run completes, signal not aborted ---")
    engine = await Zv1.create(FLOW, {"keys": {}})
    res = await engine.run({"x": 42}, 5000)
    assert res is not None, "run should return a result"
    assert engine._abort_event is not None and not engine._abort_event.is_set(), \
        "signal should NOT be set on a normal run"
    print("  ok completed normally, abort signal clean")


async def test_hung_node_is_preempted() -> None:
    print("\n--- hung node is preempted at ~timeout (not parked forever) ---")
    engine = await Zv1.create(FLOW, {"keys": {}})

    async def hang(**_kwargs):
        await asyncio.Event().wait()  # never resolves

    engine.nodes["output-data"]["process"] = hang

    timeout_ms = 600
    t0 = time.time()
    raised = False
    message = ""
    try:
        await engine.run({"x": 1}, timeout_ms)
    except Exception as e:  # noqa: BLE001
        raised = True
        message = str(e)
    elapsed = (time.time() - t0) * 1000

    assert raised, "run should raise when a node hangs past the timeout"
    assert elapsed < 3000, f"run should abort near the timeout, took {elapsed:.0f}ms"
    assert engine._abort_event.is_set(), "abort signal should have fired"
    print(f"  ok aborted in {elapsed:.0f}ms (timeout {timeout_ms}ms): {message!r}")


async def main() -> None:
    print("Testing preemptive flow timeout (python)...")
    await test_normal_run_unaffected()
    await test_hung_node_is_preempted()
    print("\nAll preemptive-timeout tests passed.")


if __name__ == "__main__":
    asyncio.run(main())

File: sdks/nodejs/tests/test.preemptive-timeout.js
/**
 * Regression test for preemptive flow timeout (AbortController).
 *
 * Before: the flow timeout only set a flag checked *between* nodes, so a single
 * hung node (e.g. a stalled LLM/HTTP call) parked the execution loop and the
 * flow could run past its timeout indefinitely.
 *
 * After: run() aborts an AbortController on timeout; node execution is raced
 * against the signal, so a hung node is cut loose at ~timeout and the run
 * rejects promptly. Normal (fast) runs are unaffected.
 *
 * Deterministic & offline: we stub a node's process to hang — no network.
 *
 * Run with: node tests/test.preemptive-timeout.js
 */
import assert from 'assert';
import Zv1 from '../src/index.js';

const flow = {
  nodes: [
    { id: 'in', type: 'input-data', settings: { key: 'x' } },
    { id: 'out', type: 'output-data', settings: { key: 'out' } },
  ],
  links: [
    { from: { node_id: 'in', port_name: 'value' }, to: { node_id: 'out', port_name: 'value' } },
  ],
};

async function testNormalRunUnaffected() {
  console.log('\n--- normal run completes, signal not aborted ---');
  const engine = await Zv1.create(flow, { keys: {} });
  const res = await engine.run({ x: 42 }, 5000);
  assert.ok(res, 'run should return a result');
  assert.strictEqual(engine.abortController.signal.aborted, false, 'signal should NOT be aborted on a normal run');
  console.log('  ✓ completed normally, abort signal clean');
}

async function testHungNodeIsPreempted() {
  console.log('\n--- hung node is preempted at ~timeout (not parked forever) ---');
  const engine = await Zv1.create(flow, { keys: {} });

  // Force the terminal node to hang forever.
  engine.nodes['output-data'].process = () => new Promise(() => {});

  const timeout = 600;
  const t0 = Date.now();
  let threw = false;
  let message = '';
  try {
    await engine.run({ x: 1 }, timeout);
  } catch (e) {
    threw = true;
    message = e.message;
  }
  const elapsed = Date.now() - t0;

  assert.ok(threw, 'run should reject when a node hangs past the timeout');
  assert.ok(elapsed < 3000, `run should abort near the timeout, but took ${elapsed}ms (was it parked?)`);
  assert.strictEqual(engine.abortController.signal.aborted, true, 'abort signal should have fired');

  // #1 regression: after a timed-out run the inherited signal must be restored,
  // so a reused engine doesn't treat this run's aborted signal as an ancestor
  // (and instantly abort) on the next run().
  assert.strictEqual(engine._inheritedSignal, null, 'top-level engine has no inherited signal');
  assert.strictEqual(engine.config.signal, engine._inheritedSignal,
    'config.signal must be restored to the inherited signal after a run');

  console.log(`  ✓ aborted in ${elapsed}ms (timeout ${timeout}ms), error: "${message}"`);
}

async function main() {
  console.log('Testing preemptive flow timeout...');
  try {
    await testNormalRunUnaffected();
    await testHungNodeIsPreempted();
    console.log('\n✅ All preemptive-timeout tests passed.');
    process.exit(0);
  } catch (err) {
    console.error('\n❌ Test failed:', err.message);
    process.exit(1);
  }
}

main();
